File: packages/service-template/templates/python/src/adapters/in/kafka_consumer.py
import os
import sys
import threading

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..")))
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", "..", "generated")))
# "in" is a Python keyword, so ports/in/ can't be dotted-imported
# (`ports.in.x` is a SyntaxError) — add its directory to sys.path and
# import it bare instead.
sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", "ports", "in")))

from confluent_kafka import Consumer
from example.v1 import example_pb2
from process_example_usecase_port_in import ProcessExampleUseCasePortIn

from core.models.example import Example


class KafkaConsumer:
    def __init__(self, usecase: ProcessExampleUseCasePortIn) -> None:
        self._usecase = usecase
        brokers = os.environ.get("KAFKA_BROKERS", "localhost:9092")
        self._topic = os.environ.get("KAFKA_CONSUME_TOPIC", "example.v1.in")
        group_id = os.environ.get("KAFKA_GROUP_ID", "__service_name__")
        self._consumer = Consumer(
            {
                "bootstrap.servers": brokers,
                "group.id": group_id,
                "auto.offset.reset": "earliest",
            }
        )
        self._consumer.subscribe([self._topic])

    def start(self, stop_event: threading.Event) -> None:
        while not stop_event.is_set():
            msg = self._consumer.poll(1.0)
            if msg is None:
                continue
            if msg.error():
                print(f"__SERVICE_NAME__: consume error: {msg.error()}", file=sys.stderr, flush=True)
                continue
            value = msg.value()
            if value is None:
                continue
            generated = example_pb2.Example()
            generated.ParseFromString(value)
            example = Example(id=generated.id, payload=generated.payload)
            self._usecase.process(example)
        self._consumer.close()
        print("__SERVICE_NAME__: consumer closed cleanly", file=sys.stderr, flush=True)

File: packages/service-template/templates/rust/src/adapters/in/kafka_consumer.rs
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::time::Duration;

use prost::Message;
use rdkafka::config::ClientConfig;
use rdkafka::consumer::{BaseConsumer, Consumer};
use rdkafka::message::Message as KafkaMessage;

use crate::core::models::example::Example;
use crate::generated::Example as GeneratedExample;
use crate::ports::port_in::process_example_usecase_port_in::ProcessExampleUseCasePortIn;

pub struct KafkaConsumer {
    consumer: BaseConsumer,
    usecase: Box<dyn ProcessExampleUseCasePortIn>,
}

impl KafkaConsumer {
    pub fn new(usecase: Box<dyn ProcessExampleUseCasePortIn>) -> Self {
        let brokers = std::env::var("KAFKA_BROKERS").unwrap_or_else(|_| "localhost:9092".to_string());
        let topic = std::env::var("KAFKA_CONSUME_TOPIC")
            .unwrap_or_else(|_| "example.v1.in".to_string());
        let group_id = std::env::var("KAFKA_GROUP_ID").unwrap_or_else(|_| "__service_name__".to_string());

        let consumer: BaseConsumer = ClientConfig::new()
            .set("bootstrap.servers", &brokers)
            .set("group.id", &group_id)
            .set("auto.offset.reset", "earliest")
            .create()
            .expect("failed to create Kafka consumer");
        consumer
            .subscribe(&[topic.as_str()])
            .expect("failed to subscribe to Kafka topic");

        Self { consumer, usecase }
    }

    pub fn start(&self, stop_flag: Arc<AtomicBool>) -> Result<(), String> {
        while !stop_flag.load(Ordering::SeqCst) {
            if let Some(result) = self.consumer.poll(Duration::from_secs(1)) {
                match result {
                    Ok(msg) => {
                        if let Some(payload) = msg.payload() {
                            let generated =
                                GeneratedExample::decode(payload).map_err(|e| e.to_string())?;
                            let example = Example {
                                id: generated.id,
                                payload: generated.payload,
                            };
                            self.usecase.process(example)?;
                        }
                    }
                    Err(e) => eprintln!("__SERVICE_NAME__: consume error: {e}"),
                }
            }
        }
        eprintln!("__SERVICE_NAME__: consumer loop exited cleanly");
        Ok(())
    }
}
